File: chapter-1-classical-mechanics/src/harmonic_oscillator.m
clear; close all;
m=1; lambda=1; dt=0.01; n_step=1000;
p=0; q=1; 
pp=zeros(n_step,1); qq=zeros(n_step,1);
for step=1:n_step
    p=p-(dt/2)*m*lambda*lambda*q; 
    q=q+(dt/m)*p;
    p=p-(dt/2)*m*lambda*lambda*q;
    pp(step,:)=p; qq(step,:)=q;
end

figure;
plot((1:n_step)*dt,qq,'linewidth',2);hold on;
plot((1:n_step)*dt,pp,'--','linewidth',2);hold on;
xlabel('time');
ylabel('position or momentum');
legend('position','momentum');
set(gca,'fontsize',16);

figure;
plot((1:n_step)*dt,0.5*m*qq.^2,'linewidth',2);hold on;
plot((1:n_step)*dt,pp.^2/m/2,'--','linewidth',2);hold on;
plot((1:n_step)*dt,0.5*m*qq.^2+pp.^2/m/2,'-.','linewidth',3);hold on;
xlabel('time');
ylabel('Energy');
legend('potential','kinetic','total');
set(gca,'fontsize',16);

figure;
plot((1:n_step)*dt,0.5*m*qq.^2+pp.^2/m/2,'-.','linewidth',3);hold on;
xlabel('time');
ylabel('Total Energy');
set(gca,'fontsize',16);

figure;
plot(qq,pp,'.','markersize',20);
xlabel('position');
ylabel('momentum');
set(gca,'fontsize',16);

File: chapter-01-classical_physics/src/harmonic_oscillator.py
import numpy as np
import matplotlib.pyplot as plt

m=1; k=1; dt=0.01; n_step=1000
v=0; x=1
v_vector = np.zeros(n_step)
x_vector = np.zeros(n_step)
for step in range(1, n_step+1):
    v = v - (dt/2) * k * x
    x = x + dt * v
    v = v - (dt/2) * k * x
    v_vector[step-1] = v
    x_vector[step-1] = x

plt.figure()
plt.plot(np.arange(1, n_step+1) * dt, x_vector, linewidth=2)
plt.plot(np.arange(1, n_step+1) * dt, v_vector, '--', linewidth=2)
plt.xlabel('time')
plt.ylabel('position or velocity')
plt.legend(['position', 'velocity'])
plt.savefig('fig-chapter-1-position_and_momentum.pdf')

potential = 0.5 * k * x_vector**2
kinetic = 0.5 * m * v_vector**2

plt.figure()
plt.plot(np.arange(1, n_step+1) * dt, potential, linewidth=2)
plt.plot(np.arange(1, n_step+1) * dt, kinetic, '--', linewidth=2)
plt.plot(np.arange(1, n_step+1) * dt, potential + kinetic, '-.', linewidth=3)
plt.xlabel('time')
plt.ylabel('Energy')
plt.legend(['potential', 'kinetic', 'total'])
plt.savefig('fig-chapter-1-energy_conservation.pdf')

plt.figure()
plt.plot(x_vector, v_vector, '.', markersize=20)
plt.xlabel('position')
plt.ylabel('momentum')
plt.savefig('fig-chapter-1-phase_space.pdf', dpi=200)
